lib/spack/spack/cmd/buildcache.py

```diff
@@ -0,0 +1,204 @@
+    create.add_argument(
+        'packages', nargs=argparse.REMAINDER,
+        help="specs of packages to create buildcache for")
+    create.set_defaults(func=createtarball)
+
+    install = subparsers.add_parser('install')
+    install.add_argument('-f', '--force', action='store_true',
+                         help="overwrite install directory if it exists.")
+    install.add_argument('-y', '--yes-to-all', action='store_true',
+                         help="answer yes to all install unsigned " +
+                              "buildcache questions")
+    install.add_argument(
+        'packages', nargs=argparse.REMAINDER,
+        help="specs of packages to install biuldache for")
+    install.set_defaults(func=installtarball)
+
+    listcache = subparsers.add_parser('list')
+    listcache.add_argument(
+        'packages', nargs=argparse.REMAINDER,
+        help="specs of packages to search for")
+    listcache.set_defaults(func=listspecs)
+
+    dlkeys = subparsers.add_parser('keys')
+    dlkeys.add_argument(
+        '-i', '--install', action='store_true',
+        help="install Keys pulled from mirror")
+    dlkeys.add_argument(
+        '-y', '--yes-to-all', action='store_true',
+        help="answer yes to all trust questions")
+    dlkeys.set_defaults(func=getkeys)
+
+
+def createtarball(args):
+    if not args.packages:
+        tty.die("build cache file creation requires at least one" +
+                " installed package argument")
+    pkgs = set(args.packages)
+    specs = set()
+    outdir = os.getcwd()
+    if args.directory:
+        outdir = args.directory
+    signkey = None
+    if args.key:
+        signkey = args.key
+    yes_to_all = False
+    force = False
+    relative = False
+    if args.yes_to_all:
+        yes_to_all = True
+    if args.force:
+        force = True
+    if args.rel:
+        relative = True
+    for pkg in pkgs:
+        for spec in spack.cmd.parse_specs(pkg, concretize=True):
+            specs.add(spec)
+            tty.msg('recursing dependencies')
+            for d, node in spec.traverse(order='post',
+                                         depth=True,
+                                         deptype=('link', 'run'),
+                                         deptype_query='run'):
+                if not node.external:
+                    tty.msg('adding dependency %s' % node.format())
+                    specs.add(node)
+    for spec in specs:
+        tty.msg('creating binary cache file for package %s ' % spec.format())
+        build_tarball(spec, outdir, force, relative, yes_to_all, signkey)
+
+
+def installtarball(args):
+    if not args.packages:
+        tty.die("build cache file installation requires" +
+                " at least one package spec argument")
+    pkgs = set(args.packages)
+    specs, links = spack.binary_distribution.get_specs()
+    matches = set()
+    for spec in specs:
+        for pkg in pkgs:
+            if re.match(re.escape(pkg), str(spec)):
```

Comment:
This is meant to cover the case where a generic expression is used, eg.
'spack buildcache install gcc'